### PR TITLE
[SDEM] Completing analysis stage design

### PR DESCRIPTION
**Pull request by GuillermoCasas**
This pull request completes an initial analysis_stage-based design of SDEM, such that both the DEM tests, the SDEM Candelier test and a simple example of a single particle settling in a cube (run by adapting a case created with the old interface and the new fluid interface).

In the following PR I intend to replace the manual Candelier test with three Kratos-style tests, which will now be possible thanks to the new design.

@RiccardoRossi @pooyan-dadvand: You can see that the new design is based on the solver, as you requested. the design was finally able to accommodate our algorithm, although unanticipated incompatibility issues could arise in the future. However, I think these are good news overall.

Unlike in FSIApplication though, the SwimmingDEM stage holds references to both the DEM stage and the FluidDynamics stage in addition to the solver's own references to the corresponding sub-solvers.

### CODE REVIEW

File: applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py [hunk 1]
```diff
@@ -323,8 +338,6 @@ def Initialize(self):
         self.MPI_results = MPI_results
         #self.mixed_model_part = self.all_model_parts.Get('MixedPart')
 
-        vars_man.ConstructListsOfVariables(self.pp)
-
         self.TransferBodyForceFromDisperseToFluid()
 
         self.FluidInitialize()
```
Comment (ipouplana) [inline applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py:343]: It should call DispersePhaseInitialize() as we discussed

File: applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py [hunk 2]
```diff
@@ -609,175 +603,82 @@ def MakeAnalyticsMeasurements(self):
...
-                    self.drag_file_output_list,
-                    self.fluid_model_part,
-                    self.time)
+    def FinalizeSolutionStep(self):
+        # printing if required
+        if self._GetSolver().CannotIgnoreFluidNow():
+            self.fluid_solution.FinalizeSolutionStep()
 
-            if self.print_counter_updated_DEM.Tick():
+        self.disperse_phase_solution.FinalizeSolutionStep()
```
Comment (ipouplana) [inline applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py:631]: Perhaps, we could call disparse_phase finalize before the fluid's

File: applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py [hunk 3]
```diff
@@ -609,175 +603,82 @@ def MakeAnalyticsMeasurements(self):
...
+        if self.debug_info_counter.Tick():
+            self.dem_volume_tool.UpdateDataAndPrint(
+                self.pp.CFD_DEM["fluid_domain_volume"].GetDouble())
 
-            self.derivative_recovery_counter.Activate(self.time > interaction_start_time)
+        super(SwimmingDEMAnalysis, self).FinalizeSolutionStep()
 
-            if self.derivative_recovery_counter.Tick():
-                self.RecoverDerivatives()
+    def RunSolutionLoop(self):
```
Comment (ipouplana) [inline applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py]: I think, this method is defined twice

File: applications/SwimmingDEMApplication/python_scripts/swimming_DEM_solver.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+        # Check for stationarity: this is useful for steady-state problems, so that
+        # the calculation stops after reaching the solution.
+        if self.stationarity_counter.Tick():
+            self.AssessStationarity()
+
+        # Compute nodal quantities to be printed that are not generated as part of the
+        # solution algorithm. For instance, the pressure gradient, which is not used for
+        # the coupling but can be of interest.
+        self.ComputePostProcessResults() # TODO: Make sure this is run only when needed
```
Comment (ipouplana) [inline applications/SwimmingDEMApplication/python_scripts/swimming_DEM_solver.py]: We already moved this into the analysis stage

File: applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py (hunk 1 above)
Comment (GuillermoCasas) [reply applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py:343]: In fact, it already does; the call happens right below 'FluidInitialize()'

File: applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py (hunk 3 above)
Comment (GuillermoCasas) [reply applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py]: true... I just checked and they are identical definitions.

File: applications/SwimmingDEMApplication/python_scripts/swimming_DEM_solver.py (hunk 4 above)
Comment (GuillermoCasas) [reply applications/SwimmingDEMApplication/python_scripts/swimming_DEM_solver.py]: Removed.

File: applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py (hunk 2 above)
Comment (GuillermoCasas) [reply applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py:631]: For the moment, I will keep this ordering. I prefer to modify the fluid (backward coupling) when all operations (FinalizeSolutionStep) have concluded there. In this way I make sure I have the last word.

**Review by ipouplana — APPROVED** [review]
Great!

File: applications/DEMApplication/python_scripts/main_script.py [hunk 5]
```diff
@@ -437,7 +437,7 @@ def RunMainTemporalLoop(self):
 
             self.AfterSolveOperations()
 
-            self.DEMFEMProcedures.MoveAllMeshes(self.all_model_parts, self.time, self.solver.dt)
+            self.solver.MoveAllMeshes(self.time, self.solver.dt)
```
Comment (RiccardoRossi) [inline applications/DEMApplication/python_scripts/main_script.py:440]: you cannot call MoveAllMeshes ... it is not a public funciton of solver

File: applications/DEMApplication/python_scripts/sphere_strategy.py [hunk 6]
```diff
@@ -319,12 +320,41 @@ def SolveSolutionStep(self):
         self.FixExternalForcesManually(time)
         (self.cplusplus_strategy).Solve()
 
-    def AdvanceInTime(self, step, time):
+    def AdvanceInTime(self, step, time, is_time_to_print = False):
         step += 1
-        time = time + self.dt
-
+        time += self.dt
+        self.UpdateTimeInModelParts(time, step, is_time_to_print)
```
Comment (RiccardoRossi) [inline applications/DEMApplication/python_scripts/sphere_strategy.py:326]: UpdateTimeInModelParts is a protected method. It should start with _ to reflect that

File: applications/DEMApplication/python_scripts/sphere_strategy.py [hunk 7]
```diff
@@ -319,12 +320,41 @@ def SolveSolutionStep(self):
...
         return step, time
 
+    def MoveAllMeshes(self, time, dt):
```
Comment (RiccardoRossi) [inline applications/DEMApplication/python_scripts/sphere_strategy.py:329]: also this is protected (or private)

if protected it should start with _
if private it should start with __

File: applications/DEMApplication/python_scripts/sphere_strategy.py [hunk 8]
```diff
@@ -319,12 +320,41 @@ def SolveSolutionStep(self):
...
+        DEM_inlet_model_part = self.all_model_parts.Get("DEMInletPart")
+        rigid_face_model_part = self.all_model_parts.Get("RigidFacePart")
+        cluster_model_part = self.all_model_parts.Get("ClusterPart")
+
+        self.mesh_motion.MoveAllMeshes(rigid_face_model_part, time, dt)
+        self.mesh_motion.MoveAllMeshes(spheres_model_part, time, dt)
+        self.mesh_motion.MoveAllMeshes(DEM_inlet_model_part, time, dt)
+        self.mesh_motion.MoveAllMeshes(cluster_model_part, time, dt)
+
+    def UpdateTimeInModelParts(self, time, step, is_time_to_print = False):
```
Comment (RiccardoRossi) [inline applications/DEMApplication/python_scripts/sphere_strategy.py:341]: initial underscore

File: applications/DEMApplication/python_scripts/sphere_strategy.py [hunk 9]
```diff
@@ -319,12 +320,41 @@ def SolveSolutionStep(self):
...
+        spheres_model_part = self.all_model_parts.Get("SpheresPart")
+        cluster_model_part = self.all_model_parts.Get("ClusterPart")
+        DEM_inlet_model_part = self.all_model_parts.Get("DEMInletPart")
+        rigid_face_model_part = self.all_model_parts.Get("RigidFacePart")
+
+        self.UpdateTimeInOneModelPart(spheres_model_part, time, step, is_time_to_print)
```
Comment (RiccardoRossi) [inline applications/DEMApplication/python_scripts/sphere_strategy.py:347]: initial underscore

File: applications/DEMApplication/python_scripts/sphere_strategy.py [hunk 10]
```diff
@@ -319,12 +320,41 @@ def SolveSolutionStep(self):
...
+        self.UpdateTimeInOneModelPart(cluster_model_part, time, step, is_time_to_print)
+        self.UpdateTimeInOneModelPart(DEM_inlet_model_part, time, step, is_time_to_print)
+        self.UpdateTimeInOneModelPart(rigid_face_model_part, time, step, is_time_to_print)
+
+    def UpdateTimeInOneModelPart(self, model_part, time, step, is_time_to_print = False):
```
Comment (RiccardoRossi) [inline applications/DEMApplication/python_scripts/sphere_strategy.py:352]: initial underscore

File: applications/DEMApplication/test_examples/basic_benchmarks/DEM_benchmarks_analysis.py [hunk 11]
```diff
@@ -164,12 +161,12 @@ def GetProblemTypeFilename(self):
 
     def BeforeSolveOperations(self, time):
```
Comment (RiccardoRossi) [inline applications/DEMApplication/test_examples/basic_benchmarks/DEM_benchmarks_analysis.py:162]: initial underscore

**Review by RiccardoRossi — DISMISSED** [review]
@GuillermoCasas you are publishing some functions which are protected or private.
please have protected functions to start with _ and private one to start with __

This is not cosmetic, and the interpreter should forbid you to call private or protected functions from outside of where they can be called (for example you cannot call them from the main script)
Even in case the interpreter does not help you, visual inspection will help you a lot if you follow this rule

File: applications/DEMApplication/python_scripts/sphere_strategy.py (hunk 10 above)
Comment (maceligueta) [reply applications/DEMApplication/python_scripts/sphere_strategy.py:352]: These methods are not newly created. They are quite old, from a time when the Analysis stage did not exist, and we had not adopted these rules yet. Changing the name of some of these methods should be done carefully, maybe with warnings for some time, and I would dedicate a specific PR for that (we don't know who is calling them from private applications).

File: applications/SwimmingDEMApplication/python_scripts/swimming_DEM_procedures.py [hunk 12]
```diff
@@ -219,11 +219,11 @@ def FindMaxNodeId(fluid_model_part):
 def FindMaxElementId(fluid_model_part):
     return max((element.Id for element in fluid_model_part.Elements))
 
-def FunctionsCalculator(pp):
-    if pp.domain_size == 2:
+def FunctionsCalculator(domain_size=3):
```
Comment (RiccardoRossi) [inline applications/SwimmingDEMApplication/python_scripts/swimming_DEM_procedures.py:222]: can't u use the default process for this?

**Review by RiccardoRossi — COMMENTED** [review]
@maceligueta ok, i can remove my block, 

nevertheless please ensure that at least new functions in stage comply with this naming, otherwise there is no way of guaranteeing that you don't call functions from the internal API

**Comment by GuillermoCasas** [comment]
Ok, I am merging although it is clear that the work is not complete. All the comments by Riccardo will be taken into account in forthcoming pull request.

**Pull request merged**